### ISSUE TITLE
inconsistent owner in Component function (product vs development build)

### ISSUE THREAD

**Issue by LiQuidProQuo** [solidjs/solid#1561]
### Describe the bug

A component function owner is different between a dev and prod build

```jsx
function MyComponent(){
     getOwner()   // <--  this is different between prod/dev builds.
                  //      in prod the owner is of the parent, in dev it is a new owner scope.
     return <></>
}
```

this inconsistency causes unexpected behavior, for example when using solid's `onCleanup` primitive 

[dev build playground](https://playground.solidjs.com/anonymous/42ddfed8-dade-4486-bae6-9b2e20647286)
![image](https://user-images.githubusercontent.com/105608035/219559861-0bbccbd1-987e-4839-bf74-9a96ab17a8b8.png)

[production build playground](https://solid-playground.netlify.app/?version=1.6.2#NobwRAdghgtgpmAXGGUCWEwBowBcCeADgsrgM4Ae2YZA9gK4BOAxiWGjIbY7gAQi9GcCABM4jXgF9eAM0a0YvADo1aAGzQiAtACsyAegDucAEYqA3EogcuPfr2ZCouOAGU0Ac2hqsvWhABhNTgoCHpCKVl5RRU6DW09CysrGXoIZlw0f14AgAs0NREACgBKfiteP0Dg0PCi0t4AXgA+cohKyuZ-OLgAOjVaDyKVZnzClRLLdqlJisE4XCZ2lSSISWSIVPTM7IBBQkIGkDn-IJCww4aWto6HbvU+gaGVKAOJqcrJWYg5oUXGdoAHjyBREvH0zSm6x+6W6fGAZFytEMvjIC1cSMMAF0mg4nC53F4oGoirhGPQ4N80bgACocOAMXD1MrXakY5FFGTEtElb5WISicTMpqtQHNECIjllABk0t4gP2EQhkkBEN8IlozHo8AguF6HgWAFFgjrcAAhfAASWKLzeYBKAEJvmBJFigA)

![image](https://user-images.githubusercontent.com/105608035/219559802-9c29a273-ef87-4fd0-a7ed-bb7a2e642121.png)

note: production build is working as designed, it is dev that alters the expectations 

### Your Example Website or App

.

### Steps to Reproduce the Bug or Issue

see playground links  for the difference between dev and prod

### Expected behavior

it is expected that dev and prod will work similarly, in regard to owner scopes.

### Platform

.

**Comment by lxsmnsyc** [comment]
If you're using the Vite plugin (or any plugin that is using Solid Refresh), this is by design. The reason is because Solid Refresh wraps the component into a memo managed by proxy HMR component. You're always going to get the owner that is provided by the memo.

**Comment by thetarnav** [comment]
Yeah, the issue is mostly about the order of `onCleanup` calls. Probably should be titled better.

**Comment by ryansolid** [comment]
I guess the fix is to exempt it as an owner. Have the parent still own everything.

**Comment by LiQuidProQuo** [comment]
@thetarnav and I looked into possible fixes, we tested this as a fix
what I did is temporarily 
![image](https://user-images.githubusercontent.com/105608035/219939369-927201c8-6e17-4ee3-af9a-5507bfff9454.png)

`isHMROwner => isDevOwner`


I did a minimal in browser patching, and it seem to correct the order with the very basic use case.
I am not sure if `onCleanup` is the only place we want/need this, but it is a start

**Comment by ryansolid** [comment]
I was playing around with something similar. At first I simply tried to hoist up the owner in `devComponent`, but obviously that would break DevTools.

Was this approach enough to get back out... because if I remember HMR also wraps things.. is it a Memo or Signal? If it is a signal that is fine. But if it is a Memo then we have another owner in here.

**Comment by LiQuidProQuo** [comment]
if we need to skip hmr owner, we can tag the owner and skip it too 
```js
if(DEV){ 
 Owner  = (Owner && Owner.isDevOwner)
                 ?  (Owner.Owner && Owner.Owner.isHMROwner)  
                      ? Owner.owner.owner       // skip Dev and Hmr Owner
                      : Owner.owner             // skip  Dev Owner
                : Owner                         // no jump
}
....

DEV && (Owner = _Owner)
```

but with HMR you have to worry about potentially not executing the delegated cleanups if the cleanup is due to an HMR refresh
and  do execute if the parent dispose is what trigger the cleanup at the lower level.
